components/range_cache_memory_engine/src/range_stats.rs

```diff
@@ -175,98 +192,228 @@ impl RangeStatsManager {
+            // expected mvcc amplification factor after cache. Make the half of it to filter
+            // the cached regions.
+            (
+                max_mvcc_amplification / f64::max(1.0, ma_mvcc_amplification_reduction / 2.0),
+                regions_to_load,
+            )
+        };
+
+        info!(
+            "IME mvcc amplification reduction filter";
+            "mvcc_amplification_to_filter" => mvcc_amplification_to_filter,
+        );
+
+        {
+            // TODO(SpadeA): remove it after it's stable
+            let debug: Vec<_> = regions_activity
+                .iter()
+                .map(|(r, s)| {
+                    format!(
+                        "region_id={}, cop={}, cop_detail={:?}, mvcc_amplification={}",
+                        r.id,
+                        s.query_stats.coprocessor,
+                        s.cop_detail,
+                        s.cop_detail.mvcc_amplification(),
+                    )
+                })
+                .collect();
+            info!(
+                "ime collect regions activities";
+                "regions" => ?debug,
+            );
+        }
+
+        let reach_stop_load = memory_controller.reached_stop_load_limit();
+        let mut regions_loaded = self.region_loaded_at.write().unwrap();
+        // Evict at most 1/10 of the regions.
+        let max_count_to_evict = usize::max(1, regions_activity.len() / 10);
+        // Use top MIN_REGION_COUNT_TO_EVICT regions next and prev to filter regions
+        // with very few next and prev. If there's less than or equal to
+        // MIN_REGION_COUNT_TO_EVICT regions, do not evict any one.
+        let regions_to_evict = if regions_activity.len() > MIN_REGION_COUNT_TO_EVICT {
+            let avg_top_next_prev = if reach_stop_load {
+                regions_activity
+                    .iter()
+                    .map(|r| r.1.cop_detail.iterated_count())
+                    .take(MIN_REGION_COUNT_TO_EVICT)
+                    .sum::<usize>()
+                    / MIN_REGION_COUNT_TO_EVICT
             } else {
-                None
-            }
-        });
-        regions_added_out.extend(added_ranges);
-        regions_removed_out.extend(removed_ranges);
+                0
+            };
+            let region_to_evict: Vec<_> = regions_activity
+                    .into_iter()
+                    .filter(|(_, r)| {
+                        if reach_stop_load {
+                            r.cop_detail.mvcc_amplification()
+                                < mvcc_amplification_to_filter
+                        } else {
+                            // In this case, memory usage is relarively low, we only evict those that should not be cached apparently.
+                            r.cop_detail.mvcc_amplification()
+                                <= self.mvcc_amplification_threshold as f64 / MVCC_AMPLIFICATION_FILTER_FACTOR
```

Comment:
`mvcc_amplification_threshold`'s default is 10 and MVCC_AMPLIFICATION_FILTER_FACTOR is also 10, then seem by default, this compare will always be false? 10 is too big for `MVCC_AMPLIFICATION_FILTER_FACTOR` here?

Reply:
I changed mvcc_amplification_threshold to 100 in this PR as for region with mvcc problem, it can be easy to go to serverial hundreds or thousands.